module/spock/spock_verifier.go

```diff
@@ -0,0 +1,154 @@
+package spock
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// Verifier provides functionality to verify SPoCK proofs
+type Verifier struct {
+	// state is used to query identities at a blockId to get StakingPublicKey
+	protocolState protocol.ReadOnlyState
+
+	// map of receipts by result ID that do not have matching SPoCK proofs
+	// For instance, if there are 5 receipts that have 5 SpockSets, say, Spockset1, Spockset2, Spockset3, Spockset4, and Spockset5.
+	// If Spockset1 and Spockset2 are for ER1, and they match with each other;
+	// And Spockset3, Spockset4, Spockset5 are for ER2. and they don't match with each other.
+	// Then there are 2 category, the first category has bucket: ER1_Spockset1, and the second category has buckets: ER2_Spockset3, ER2_Spockset4, and ER2_Spockset5.
+	// When we receive each approval, we will first check which category it should go, and find all the buckets to try matching against.
+	receipts map[flow.Identifier][]*flow.ExecutionReceipt
+}
+
+// NewVerifier creates a new spock verifier
+func NewVerifier(state protocol.ReadOnlyState) *Verifier {
+	return &Verifier{
+		protocolState: state,
+		receipts:      make(map[flow.Identifier][]*flow.ExecutionReceipt),
+	}
+}
+
+// AddReceipt adds a receipt into map if the SPoCK proofs do not match any other receipt
+func (v *Verifier) AddReceipt(receipt *flow.ExecutionReceipt) error {
+	resultID := receipt.ExecutionResult.ID()
+
+	// if receipts result id does not exist in map, create an array with the receipt
+	_, ok := v.receipts[resultID]
+	if !ok {
+		receipts := make([]*flow.ExecutionReceipt, 0)
+		receipts = append(receipts, receipt)
+		v.receipts[resultID] = receipts
+		return nil
+	}
+
+	// check if candidate receipt SPoCK proofs match proofs in map
+	matched, err := v.matchReceipt(receipt)
+	if err != nil {
+		return fmt.Errorf("could not match receipt: %w", err)
+	}
+
+	// if not matched add to receipt array
+	if !matched {
+		v.receipts[resultID] = append(v.receipts[resultID], receipt)
+	}
+
+	// if matched we do nothing (strong-transitive property of the SPoCK scheme)
+
+	return nil
+}
+
+// ClearReceipts clears all receipts for a specific resultID
+func (v *Verifier) ClearReceipts(resultID flow.Identifier) bool {
+
+	// check if entry exists
+	_, ok := v.receipts[resultID]
+	if !ok {
+		return false
+	}
+
+	// clear receipts
+	delete(v.receipts, resultID)
+
+	return true
+}
+
+// VerifyApproval verifies an approval with all the distinct receipts for the approvals
+// result id and returns true if spocks match else false
+func (v *Verifier) VerifyApproval(approval *flow.ResultApproval) (bool, error) {
```

Comment:
When verifying an approval, if the approval matches with a receipt, we need to return the matched receipt. Returning a bool might not be enough

Consider this case:
There are two receipts X and Y having the same result. `X` has valid Spork, `Y` has `invalid` Spork. Now the result has two chunks, and we require each chunks to receive any 2 approvals from 3 approvers (`J, K, L`) to seal the result. 
Now, if `J` approves `X`, and `K` , who is malicious, approves `Y`. Meaning, `J`’s spork matches with `X`’s spork, and `K`’s invalid spork matches with `Y`’s invalid spork, then it seems there are 2 approvals approving the same result, but actually they are approving different receipts. If we don't distinguish, we might seal a result that haven't received enough valid approvals.

Reply:
We could return the receipt which each approval matched and collect them on the consensus matching engine and also keep track about the number of approval matches. Then once we go through all the approvals for a chunk we should check if the highest matching receipt and use that as the total number of valid approvals. What do you think about this approach?

Reply:
Yes, I think this function needs to return the receipt that has the matched Spock with the approval. 

And yes, the threshold should be checked with the highest matching receipt.

Reply:
So as it is now, I've made the consensus matching engine try and verify an approval as they come in and cache them for verification later when `checkSealing` gets called. It could also not match since the receipt might have been received later so we only cache if it has a match.

When matching a receipt the time complexity is `U * C` where `U` is the number of receipts with non-matching SPoCK proofs and `C` is the number of chunks per Receipt's Execution Result. When matching an approval SPoCK to a receipt the time complexity is `U`.

Reply:
There way we look for sealable results should definitely change to be more optimized, though that logic wasn't introduced by this PR and it doesn't have to be fixed in this PR/ticket. 

I think the issue is that for every new approval a consensus node receives, it looks at the mempool of all results and checks which ones can be sealed by verifying all the SPoCK proofs for all the chunks again (if my understanding of the code is correct). This call here should be updated: https://github.com/onflow/flow-go/blob/danu/1325-verify-spock-proofs/engine/consensus/matching/engine.go#L379 to look only at the result concerned by the new approval. If there is a result that could be sealed/updated it would only be the one linked to the received approval. 

Is there already a ticket to update this logic?

Reply:
@tarakby , I just created the issue here https://github.com/dapperlabs/flow-go/issues/4933

I am not sure how high priority this should be. The problem is that we don't have a full picture of how the sealing logic ties together yet in the context of fork-awareness.

For example, one problem with the solution you mentionned is that there are other external conditions that can make a result sealable. If a result has received all the approvals but we don't know its block yet, it will never be sealed, unless we rerun the checks. In the future, we will implement some kind of subscription model to notify the matching engine when blocks or results are added to the chain.. but not 100% sure what that will look like. 

In the meantime, I have submitted a change with a more efficient counting of signatures (not SPoCK proofs), whereby we dont recheck approvals eachtime. https://github.com/onflow/flow-go/pull/75

By the way, did you finish implementing the BLS aggregated signatures? We have this TODO in our code to switch over to BLS when it's ready: https://github.com/onflow/flow-go/blob/f6bb825fd5c1fbab48df2a10d09c31838cae05d7/model/flow/aggregated_signature.go#L7

Reply:
Thanks for creating the ticket and tracking the issue. 

I think we are aligned, what I meant is that *if* there is a result to be updated, it would be only about the result concerned by the new approval (that goes from just updating the existing count of a chunk approvals, to sealing the entire result). 

Thanks, I'll look at the changes in #75. 

BLS aggregation is ready. However BLS aggregation doesn't allow the methods implemented in this file: https://github.com/onflow/flow-go/blob/f6bb825fd5c1fbab48df2a10d09c31838cae05d7/model/flow/aggregated_signature.go#L7. If the verification logic is relying on all these methods, then aggregation might not be possible. We could sync about this if you have questions. 
There is also another module that is using an aggregation API (used in consensus) that we could re-use: https://github.com/onflow/flow-go/blob/f6bb825fd5c1fbab48df2a10d09c31838cae05d7/module/signature/aggregation.go